DIPs/1NNN-AP.md

```diff
@@ -0,0 +1,172 @@
+
+Java allows omitting an enum type in switch-case statements: [https://docs.oracle.com/javase/tutorial/java/javaOO/enum.html](https://docs.oracle.com/javase/tutorial/java/javaOO/enum.html)
+
+A suggestion for a similar feature in Rust, but with a different approach: [https://internals.rust-lang.org/t/enum-path-inference-with-variant-syntax/16851](https://internals.rust-lang.org/t/enum-path-inference-with-variant-syntax/16851)
+
+## Description
+Due to `.enumMember` being the same syntax as the [module scope operator](https://dlang.org/spec/module.html#module_scope_operators),
+I propose that ETI should use the syntax `$enumMember`. I'm open to discussing alternatives,
+but I believe that my proposed syntax represents the best compromise between convenience and compatibility.
+
+#### 1. Assignment statements.
+Assigning to a variable with a known type should allow ETI.
+```d
+enum A{ a,b,c,d; }
+
+struct B{ A one, two; }
+
+void main(){
+    A    myA1 = $b; //myA1 = A.b
+    auto myA2 = $b; //error, we don't know the type of "$b"!
+    
+    B myB;
+    myB.one = $c; //myB.one = A.c
+    myB.two = $d; //myB.two = A.d
+}
+```
+
+#### 2. Return statements.
+If a function has an explicit enum return type, its return statement(s)
+should allow ETI.
+```d
+enum A{ a,b,c,d; }
+
+A myFunc(){
+    return $c; //returns A.c
+}
+
+auto myBrokenFunc(){
+    return $c; //error, we don't know the type of "$c"!
+}
+```
+
+#### 3. Parameters.
+Most parameters require explicit typing beforehand, and thus should
+always allow ETI. The exception is `alias` template parameters, which
+cannot use ETI.
+```d
+enum A{ a,b,c,d; }
+
+struct B{ A one, two; }
+
+void myFunc(A param){}
+void myTempFunc(T)(T param){}
+
+void main(){
+    B    myB1 = {one:$a, two:$b};
+    auto myB2 = B($a, $b);
+    
+    myFunc($a);
+    myTempFunc!A($a);
+    myTempFunc($a); //error, can't infer a type to instantiate template with from "$a"
+}
+```
+
+#### 4. Switch-case statements.
+Switch-case statements should have a special syntax that works the same way as `switch(myVar) with(typeof(myVar))`.
+```d
+enum WordLetterOfTheDay{ a,b,c,d/*...*/; }
+
+void main(){
+    auto letterToday = WordLetterOfTheDay.b;
+    
+    import std.stdio;
+    switch(letterToday){
+        case a:
+            writeln("Apple");
+            break;
+        case b:
+            writeln("Bicycle");
+            break;
+        case c:
+            writeln("Caterpillar");
+            break;
+        case d:
+            writeln("Didgeridoo");
+            break;
+        /*...*/
+    }
+}
+```
+
+#### 5. Array literals.
+When an array literal has a specified type, ETI should always be allowed(1).
+
+
+
+When an array literal has an ambiguous type, I propose that if an enum type is explicitly
+used for the first array item, ETI should be allowed in the remainder of the array.
+```d
+enum A{ a,b,c,d; }
+
+A[4] x = [$a, $b, $c, $d];  //(1)
+auto y = [A.a, $b, $c, $d]; //(2)
```

Comment:
If the literal has a different inferred common type, does this rule still apply? E.g. `[A.a, 7, $b]`.

Reply:
> If the literal would have a different inferred common type, does this rule still apply? E.g. `[A.a, 7, $b]`.

Apparently that would turn the whole array into an array of ints? That's some weird cast behaviour, but I don't see why it would affect this rule (e.g. by creating ambiguity or breaking old syntax).

Reply:
To clarify, I think that `auto y = [A.a, 7, $b];` would turn into `[0, 7, 1]`.